### PR TITLE
Issue #4763: fix empty statement for comment under package

### PR DESCRIPTION
**Pull request by strkkk**
#4763

Reg report: https://strkkk.github.io/4763/

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 1]
```diff
@@ -486,6 +482,11 @@ private void processPackage(DetailAST ast, DetailAST nextToken) {
         if (!hasEmptyLineAfter(ast)) {
             log(nextToken.getLineNo(), MSG_SHOULD_BE_SEPARATED, nextToken.getText());
         }
+        // workaround of antlr bug, it puts comments into type definition modifiers
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: it is not a bug, it is by design.
please change comment or remove it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 2]
```diff
@@ -474,7 +470,7 @@ private boolean hasMultipleLinesBefore(DetailAST ast) {
      */
     private void processPackage(DetailAST ast, DetailAST nextToken) {
         if (ast.getLineNo() > 1 && !hasEmptyLineBefore(ast)) {
-            if (getFileContents().getFileName().endsWith("package-info.java")) {
+            if (getFileContents().inPackageInfo()) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: would be awesome to move to separate commit (first commit of this PR), as it is not related issue

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 3]
```diff
@@ -458,13 +459,8 @@ private static boolean isClassMemberBlock(int astType) {
...
     private boolean hasMultipleLinesBefore(DetailAST ast) {
-        boolean result = false;
-        if ((ast.getType() != TokenTypes.VARIABLE_DEF
-            || isTypeField(ast))
-                && hasNotAllowedTwoEmptyLinesBefore(ast)) {
-            result = true;
-        }
-        return result;
+        return (ast.getType() != TokenTypes.VARIABLE_DEF || isTypeField(ast))
+                    && hasNotAllowedTwoEmptyLinesBefore(ast);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: if this is just refactoring, please move to separate commit (without reference  to issue in commit name)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 4]
```diff
@@ -623,6 +624,19 @@ private boolean hasEmptyLineAfter(DetailAST token) {
...
+     * Finds first modifier of the token.
+     *
+     * @param token token to check
+     * @return first modifier or {@code null} if there is no modifier
+     */
+    private static DetailAST findFirstModifier(DetailAST token) {
+        return Optional.ofNullable(token)
+            .map(sibling -> sibling.findFirstToken(TokenTypes.MODIFIERS))
+            .map(DetailAST::getFirstChild)
+            .orElse(null);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: does it make sense to avoid null return ?
as we become to use more and more Optional.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 5]
```diff
@@ -683,13 +697,9 @@ private boolean isCommentInBeginningOfLine(DetailAST comment) {
...
     private static boolean isPrecededByJavadoc(DetailAST token) {
-        boolean result = false;
         final DetailAST previous = token.getPreviousSibling();
-        if (previous.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
-                && JavadocUtil.isJavadocComment(previous.getFirstChild().getText())) {
-            result = true;
-        }
-        return result;
+        return previous.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
+                && JavadocUtil.isJavadocComment(previous.getFirstChild().getText());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: please never mix functional fix in refactoring in same commit.

I would do:
```
result = previous.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
                && JavadocUtil.isJavadocComment(previous.getFirstChild().getText());
return result;
```
as it would ease debugging, and will easily be optimized by compiler. I am old-school.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorBlockCommentSeparatedFromPackage.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace.emptylineseparator;
+
+/*
+* My important message
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorBlockCommentSeparatedFromPackage.java]: wan we use our approach in Inputs to use comments as place for violation ?  `//warn` `/* warn */`
with definition of config above class.

in all other inputs too.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 5 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: @romani there are plenty of methods like this (return expression without unnecessary variable), even in this class almost all other methods do it in same way, e.g. this one (2 lines below current)
```
private static boolean isComment(DetailAST ast) {
        return ast.getType() == TokenTypes.SINGLE_LINE_COMMENT
                   || ast.getType() == TokenTypes.BLOCK_COMMENT_BEGIN;
    }
```

What is wrong with debugging?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 3 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 4 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorBlockCommentSeparatedFromPackage.java (hunk 6 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorBlockCommentSeparatedFromPackage.java]: done

**Comment by strkkk** [comment]
Looks like it fixed not only cases with comment, but with class definition directly under package as well, e.g.
https://strkkk.github.io/4763/checkstyle/index.html#A3

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 3 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 7]
```diff
@@ -481,6 +482,9 @@ private void processPackage(DetailAST ast, DetailAST nextToken) {
         if (!hasEmptyLineAfter(ast)) {
             log(nextToken.getLineNo(), MSG_SHOULD_BE_SEPARATED, nextToken.getText());
         }
+        findFirstModifier(nextToken)
+            .filter(mod -> mod.getLineNo() == ast.getLineNo() + 1)
+            .ifPresent(mod -> log(mod.getLineNo(), MSG_SHOULD_BE_SEPARATED, mod.getText()));
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: probably it is good time to start discussion and agree on how far we go with functional style.
In general I am in ok, but here is special case of `log` in lambda.
In most cases I demand `log` to be placed in `visitToken` with some exceptions when Check is complicated, or it was like this for long time, and we do not have time for refactor.
I think that we should not place `log` into lambda, because we lost ability of coverage/pitest to automatically show us such line is covered. Functional style is very good style when engineering quality if measured by coverage, single test can cover all code :), but this is not a reason why we use coverage. Coverage helps us to grow Input code base with cases. Contributors always try to provide as minimum tests as possible.

If would do:
```
if (variable.isPresent()) {
  log()
}
``` 

@rnveach , @pbludov , please vote of your preferences.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: 1) I do not like long and complicated expressions.
2) I like single exit point concept in methods (it force to organize code well, we benefited a lot after enforcement, even process was also not welcomed at  the beginning), when you have code like this you put a break point to `return result` and you easily have final value, if parts of expressions do not do side effects, you can re-evaluate parts of expression to get to know which part of it result in unexpected result.
3) simple expressions are growing in more complicated , and I not always have ability to catch it during review. I am not ready to design Check on this, as there are a lot of nuances in objective term `complicated expression`.

Lets vote on this.
@rnveach  @pbludov , please vote.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 7 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > I think that we should not place `log` into lambda, because we lost ability of coverage/pitest to automatically show us such line is covered

I asked this question some time ago, googled a little and tried it myself. And I haven't seen any problems with pitest and lambdas.
Can you share an example where lambda hides coverage?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 5 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: 1. how come that such expression became complicated? I can agree if there is some non-obvious things and variable name can clarify it, but create useless `result` variable is not the case.
2. I do not see any conflicts on this. You can evaluate expression if you put breakpoint. Conditional breakpoints solve this problem.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 7 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: If something hurts pitest or coverage, I too would like to avoid it. If we know something will break pitest, we should write a check to avoid it.
@strkkk Can you post pitest report for your changes?

My own personal opinion is I don't like Java 8+ functionality anyways. I would much rather see everything written it out as I don't find lambdas very clear to read.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: pitest and jacoco coverage will be green and 100% covered.
pitest does not have such mutations to substitude lambda to somethig else.
jacoco it is not smart enough for now to detect actual lambda body execution.
BUT it only mean that particular line was executed by some tests. Pitest will not do any mutation as far as I remember, so it will be green.

Functional style code is very easy to cover, as all IFs (`ifPresent`) are chained in single statement, and jacoco consider them as single statement. Any test will cover it.
Well known cheating of pitest mutations is to substitute all "+-" with `add/remove` methods. Amount of mutations will be seriously reduced, so easy to pass.

Java 8+ functionality is ok, it ease code in some cases, but only if we know that do coverage of all(very most) branches it imply.
Fight for coverage, removed a lot of nonsense code and heavily forced us to extend Input code base. Idea of almost complete ban of pure UTs is also come from idea that we do not care about coverage number, we want to have Inputs that cover code and all cases (functional test).

Our CI will force contributor to provide heavy testing if code is old-styled-java(imperative), it was good time for maintainers during code review as you now that each line/branch was covered by Input. 

So even we make our code modern in some parts, I would rather keep at least `logs` calls out of such single-long statement(long chaining of calls). I am a bit paranoid probably.

@pbludov , please share your ideas on this topic.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > My own personal opinion is I don't like Java 8+ functionality anyways. I would much rather see everything written it out as I don't find lambdas very clear to read

Not about lambdas, but optional style is a best way to dig down 2+ nullable levels down in  AST, because it eliminates useless null checks.

> Pitest will not do any mutation as far as I remember, so it will be green.

Pitest mutates conditions, even in lambdas.
Ok, I can replace ifPresent with explicit `isPresent` call

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > Ok, I can replace ifPresent with explicit isPresent call

Yes, it will shift the focus in the right direction.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 8]
```diff
@@ -623,6 +622,18 @@ private boolean hasEmptyLineAfter(DetailAST token) {
...
 
+    /**
+     * Finds first modifier of the token.
+     *
+     * @param token token to check
+     * @return first modifier of the token
+     */
+    private static Optional<DetailAST> findFirstModifier(DetailAST token) {
+        return Optional.ofNullable(token)
+            .map(sibling -> sibling.findFirstToken(TokenTypes.MODIFIERS))
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: Please explain why the name of this variable is `sibling`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 5 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: I'm agree with @strkkk. The variable `result` is redundant here.
Now the code is strictly sequential and more readable (for me).
Variables are needed when there are *real* branches.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > how come that such expression became complicated?

in future, there will be some additional request for checking in this condition and next contributor will simply extend expression with extra code

> You can evaluate expression if you put breakpoint.

and now please count a time, you put break point and now you try to accurately select whole expression in IDE to evaluate it. Compare it with simply hovering it over "result" variable. 

> Now the code is strictly sequential and more readable (for me).

with extra variable it it also sequential.

> Variables are needed when there are real branches.

I do not understand this. Variables help to organize code and keep some naming context.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 8 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: just garbage left after refactoring, changed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 7 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: changed.

Now Idea complains that this can be changed to use functional style :)

**Comment by strkkk** [comment]
> please add it to test Inputs. All unexpected/interesting cases from diff regression should be copied to Inputs. Eventually, we should not rely of diff report that much for testing.

As I saw, it was a duplicated violation. Such cases already can be detected, but there was no UT for it, I added one.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 5 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: 1. We can make some limit (if there are several conditions, it should be extracted to variables) but it will definitely more than 2, as we have here.
2. I don't see major issue in it.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > 1.

 Limit will not work, as there is required to detect "the same" pattern between parts of expression. It is a bit more complicated, I am not ready to provide all rules for such future Check. Expression with 2 items is not good example to discuss over. I just try to warn you that I am not ok to refactor whole project to make it single statement "return" with whole logic inlined.

> 2 

All Checkstyle violations are not a major issue. We focus on issues that economize minutes and seconds of user during reading, and even during debug (it is also a phase where engineers spend a lot of time). Where are areas where we still can improve(even in seconds, less actions, less distractions, less scrolling) reading experience of code. I am ok to inline variable definition by expression of IF, but I still see convenience to keep SINGLE exit point as "single" variable return (but I agree there are simple cases).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 7 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: you already should know that static analysis is full of false-positives and exceptions of rules. It is better to suppress violation if we know a good reason.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 9]
```diff
@@ -478,7 +479,14 @@ private void processPackage(DetailAST ast, DetailAST nextToken) {
                 log(ast.getLineNo(), MSG_SHOULD_BE_SEPARATED, ast.getText());
             }
         }
-        if (!hasEmptyLineAfter(ast)) {
+        if (hasEmptyLineAfter(ast)) {
+            final Optional<DetailAST> modifier = findFirstModifier(nextToken)
+                .filter(mod -> mod.getLineNo() == ast.getLineNo() + 1);
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: it is possible, is `findComment`as method name ok for you?

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: if method doe only this - sure . it will be much better to read code and completely match to issue you resolving.

**Comment by romani** [comment]
after this final change, please regenerate diff report, to make sure nothing was lost during refactoring and next review will be quick.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 10]
```diff
@@ -478,7 +479,13 @@ private void processPackage(DetailAST ast, DetailAST nextToken) {
                 log(ast.getLineNo(), MSG_SHOULD_BE_SEPARATED, ast.getText());
             }
         }
-        if (!hasEmptyLineAfter(ast)) {
+        if (hasEmptyLineAfter(ast)) {
+            final Optional<DetailAST> comment = findComment(ast);
+            if (comment.isPresent()) {
+                log(comment.get().getLineNo(), MSG_SHOULD_BE_SEPARATED, comment.get().getText());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: this code become a non-sense for reader:
```
if (hasEmptyLineAfter(ast)) { 
  ... 
  log(comment.get().getLineNo(), MSG_SHOULD_BE_SEPARATE
}
```

we need to update `hasEmptyLineAfter` with comments find.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: you cannot do that. This method is used in other methods. Also, it will log on ast but it should log on comment token.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: we shuould do refactoring on names, code should be read as English text and NO cheating.
can you propose something ?

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: Something like 
```
Optional<DetailAST> comment = findComment(ast)
if (comment.isPresent() log comment
else if(!hasEmptyLineAfter(ast) log next sibling
```
?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: much better !
please do not forget to regenerate diff report.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: changed, please take a look.

Report will be later

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 8 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: FYI I changed it back to sibling because code changed

**Comment by romani** [comment]
@strkkk , please address my comment above, regenerate diff report, to let me finish review.

**Comment by strkkk** [comment]
@romani I generated it, but after refactoring there are some extra violations, I am checking it.

**Comment by romani** [comment]
Ok, please do not forget to add them to inputs

**Comment by strkkk** [comment]
It looks like this changes affected detection for type definitions which is very weird.
I assume there was already some other bug before.
I will reopen it when it will be fixed
